packages/pds/src/api/com/atproto/server/createAccount.ts

```diff
@@ -158,3 +139,102 @@ export const ensureCodeIsAvailable = async (
+  ctx: AppContext,
+  input: CreateAccountInput,
+): Promise<string> => {
+  try {
+    const handle = ident.normalizeAndEnsureValidHandle(input.handle)
+    ident.ensureHandleServiceConstraints(handle, ctx.cfg.availableUserDomains)
+    return handle
+  } catch (err) {
+    if (err instanceof ident.InvalidHandleError) {
+      throw new InvalidRequestError(err.message, 'InvalidHandle')
+    } else if (err instanceof ident.ReservedHandleError) {
+      throw new InvalidRequestError(err.message, 'HandleNotAvailable')
+    } else if (err instanceof ident.UnsupportedDomainError) {
+      if (input.did === undefined) {
+        throw new InvalidRequestError(err.message, 'UnsupportedDomain')
+      }
+      const resolvedHandleDid = await resolveExternalHandle(
+        ctx.cfg.scheme,
+        input.handle,
+      )
+      if (input.did !== resolvedHandleDid) {
+        throw new InvalidRequestError('External handle did not resolve to DID')
+      }
+    }
+    throw err
+  }
+}
+
+const getDidAndPlcOp = async (
+  ctx: AppContext,
+  handle: string,
+  input: CreateAccountInput,
+): Promise<{
+  did: string
+  plcOp: plc.Operation | null
+}> => {
+  // if the user is not bringing a DID, then we format a create op for PLC
+  // but we don't send until we ensure the username & email are available
+  if (!input.did) {
+    const rotationKeys = [ctx.cfg.recoveryKey, ctx.plcRotationKey.did()]
+    if (input.recoveryKey) {
+      rotationKeys.unshift(input.recoveryKey)
+    }
+    const plcCreate = await plc.createOp({
+      signingKey: ctx.repoSigningKey.did(),
+      rotationKeys,
+      handle,
+      pds: ctx.cfg.publicUrl,
+      signer: ctx.plcRotationKey,
+    })
+    return {
+      did: plcCreate.did,
+      plcOp: plcCreate.op,
+    }
+  }
+
+  // if the user is bringing their own did:
+  // resolve the user's did doc data, including rotationKeys if did:plc
+  // determine if we have the capability to make changes to their DID
+  let atpData: AtprotoData
+  try {
+    atpData = await ctx.didResolver.resolveAtprotoData(input.did)
+  } catch (err) {
+    throw new InvalidRequestError(
+      `could not resolve valid DID document :${input.did}`,
+      'UnresolvableDid',
+    )
+  }
+  if (atpData.handle !== handle) {
+    throw new InvalidRequestError(
+      'provided handle does not match DID document handle',
+      'IncompatibleDidDoc',
+    )
+  } else if (atpData.pds !== ctx.cfg.publicUrl) {
+    throw new InvalidRequestError(
+      'DID document pds endpoint does not match service endpoint',
+      'IncompatibleDidDoc',
+    )
+  } else if (atpData.signingKey !== ctx.repoSigningKey.did()) {
+    throw new InvalidRequestError(
+      'DID document signing key does not match service signing key',
+      'IncompatibleDidDoc',
+    )
+  }
+
+  if (input.did.startsWith('did:plc')) {
+    const data = await ctx.plcClient.getDocumentData(input.did)
+    if (!data.rotationKeys.includes(ctx.plcRotationKey.did())) {
```

Comment:
This may not be a good idea.
If the reason a user is bringing their own did is that they don't want the PDS to be able to update their did:plc this will prevent them from using this PDS.

There is a real use case of I want to manage my own DID and just use the PDS to manage the repo.

Reply:
Yeah we will loosen this in the future, but I think it's important for now

If the PDS does not have the ability to rotate documents, then
- it cannot help out with handle changes
- if it rotates it's repo signing key, it can no longer publish records on a repo's behalf

This should be allowed at some point, but for now, for usability's sake, we constrain the provided DID